File: src/BestTimeToBuyAndSellStock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock
// T: O(n)
// S: O(1)

public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) {
            return 0;
        }

        int buyPrice = prices[0];
        int profit = 0;
        for (int index = 1 ; index < prices.length ; index++) {
            profit = Math.max(profit, prices[index] - buyPrice);
            buyPrice = Math.min(buyPrice, prices[index]);
        }
        return profit;
    }
}

File: python/best_time_to_buy_and_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit, buy_price = 0, prices[0]
        for price in prices:
            max_profit = max(max_profit, price - buy_price)
            buy_price = min(buy_price, price)
        return max_profit
